v4/supervisor.go

```diff
@@ -0,0 +1,956 @@
+
+	if sentControlErr != nil {
+		return sentControlErr
+	}
+
+	select {
+	case <-notificationC:
+		// normal case; the service is terminated.
+		return nil
+
+	// This occurs if the entire supervisor ends without the service
+	// having terminated, and includes the timeout the supervisor
+	// itself waited before closing the liveness channel.
+	case <-s.ctx.Done():
+		return ErrTimeout
+
+	// The local timeout.
+	case <-timeoutC:
+		return ErrTimeout
+	}
+}
+
+/*
+
+Services returns a []Service containing a snapshot of the services this
+Supervisor is managing.
+
+*/
+func (s *Supervisor) Services() []Service {
+	ls := listServices{make(chan []Service)}
+
+	if s.sendControl(ls) == nil {
+		return <-ls.c
+	}
+	return nil
+}
+
+var currentSupervisorIDL sync.Mutex
+var currentSupervisorID uint32
+
+func nextSupervisorID() supervisorID {
+	currentSupervisorIDL.Lock()
+	defer currentSupervisorIDL.Unlock()
+	currentSupervisorID++
+	return supervisorID(currentSupervisorID)
+}
+
+// ServiceToken is an opaque identifier that can be used to terminate a service that
+// has been Add()ed to a Supervisor.
+type ServiceToken struct {
+	id uint64
+}
+
+// An UnstoppedService is the component member of an
+// UnstoppedServiceReport.
+//
+// The SupervisorPath is the path down the supervisor tree to the given
+// service.
+type UnstoppedService struct {
+	SupervisorPath []*Supervisor
+	Service        Service
+	Name           string
+	ServiceToken   ServiceToken
+}
+
+// An UnstoppedServiceReport will be returned by StopWithReport, reporting
+// which services failed to stop.
+type UnstoppedServiceReport []UnstoppedService
+
+type serviceWithName struct {
+	Service Service
+	name    string
+}
+
+// Jitter returns the sum of the input duration and a random jitter.  It is
+// compatible with the jitter functions in github.com/lthibault/jitterbug.
+type Jitter interface {
+	Jitter(time.Duration) time.Duration
+}
+
+// NoJitter does not apply any jitter to the input duration
+type NoJitter struct{}
+
+// Jitter leaves the input duration d unchanged.
+func (NoJitter) Jitter(d time.Duration) time.Duration { return d }
+
+// DefaultJitter is the jitter function that is applied when spec.BackoffJitter
+// is set to nil.
+type DefaultJitter struct {
+	rand *rand.Rand
+}
+
+// Jitter will jitter the backoff time by uniformly distributing it into
+// the range [FailureBackoff, 1.5 * FailureBackoff).
+func (dj *DefaultJitter) Jitter(d time.Duration) time.Duration {
+	// this is only called by the core supervisor loop, so it is
+	// single-thread safe.
+	if dj.rand == nil {
+		dj.rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	}
+	jitter := dj.rand.Float64() / 2
+	return d + time.Duration(float64(d)*jitter)
+}
+
+type (
+	// BadStopLogger is called when a service fails to properly stop
+	BadStopLogger func(*Supervisor, Service, string)
+
+	// FailureLogger is called when a service fails
+	FailureLogger func(
```

Comment:
I quite often find myself writing log handlers to use github.com/lthibault/log as a logging backend. It’s quite tedious given the large arity of these functions. Consider passing in a struct instead.

Reply:
The more I've thought about it, the more sensible that seems to be.

Let's go ahead and do it.